Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp

```diff
@@ -140,13 +165,113 @@ void UnifiedTextReplacementController::textReplacementSessionDidReceiveReplaceme
         auto newResolvedRange = resolveCharacterRange(*sessionRange, newRangeWithOffset);
 
-        auto markerData = DocumentMarker::UnifiedTextReplacementData { replacementData.originalString.string, replacementData.uuid, DocumentMarker::UnifiedTextReplacementData::ApplyIndicator::Yes };
+        auto markerData = DocumentMarker::UnifiedTextReplacementData { replacementData.originalString.string, replacementData.uuid, DocumentMarker::UnifiedTextReplacementData::State::Pending };
         addMarker(resolvedRange, DocumentMarker::Type::UnifiedTextReplacement, markerData);
 
         additionalOffset += replacementData.replacement.length() - replacementData.originalRange.length;
     }
 }
 
+void UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement(const WTF::UUID& uuid, WebTextReplacementData::State state, const WebTextReplacementData& replacement, const WebUnifiedTextReplacementContextData& context)
+{
+    RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement (%s) [new state: %hhu, replacement: %s]", uuid.toString().utf8().data(), enumToUnderlyingType(state), replacement.uuid.toString().utf8().data());
+
+    if (!m_webPage) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    auto* corePage = m_webPage->corePage();
+    if (!corePage) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    Ref frame = CheckedRef(corePage->focusController())->focusedOrMainFrame();
+    RefPtr document = frame->document();
+    if (!document) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    if (state != WebTextReplacementData::State::Committed && state != WebTextReplacementData::State::Reverted)
+        return;
+
+    auto nodeAndMarker = findReplacementMarkerByUUID(document.get(), replacement.uuid);
+    if (!nodeAndMarker) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
+    auto& [node, marker] = *nodeAndMarker;
+
+    auto rangeToReplace = makeSimpleRange(node, marker);
+
+    auto oldData = std::get<DocumentMarker::UnifiedTextReplacementData>(marker.data());
+
+    auto offsetRange = OffsetRange { marker.startOffset(), marker.endOffset() };
+    document->markers().removeMarkers(node, offsetRange, { DocumentMarker::Type::UnifiedTextReplacement });
+
+    auto [newText, newState] = [&]() -> std::tuple<String, DocumentMarker::UnifiedTextReplacementData::State> {
+        switch (state) {
+        case WebTextReplacementData::State::Committed:
+            return std::make_tuple(replacement.replacement, DocumentMarker::UnifiedTextReplacementData::State::Committed);
+
+        case WebTextReplacementData::State::Reverted:
+            return std::make_tuple(oldData.originalText, DocumentMarker::UnifiedTextReplacementData::State::Reverted);
+
+        default:
+            RELEASE_ASSERT_NOT_REACHED();
+        }
+    }();
+
+    replaceTextInRange(frame.get(), rangeToReplace, newText);
```

Comment:
Just to confirm, this remove the old marker right?

Reply:
it actually probably does, but I already explicitly remove the markers anyways on line 213